middlewares/authzCourseInstance.js

```diff
@@ -86,15 +86,52 @@ module.exports = function(req, res, next) {
+                    // example, the user dropdown in the navbar to be hidden. It is likely that a better
+                    // solution can be found when this code is all rewritten.
+                    delete res.locals.authz_data;
+                    res.locals.effective_user_has_no_access_to_course_instance = true;
 
-                _.assign(res.locals.authz_data, result.rows[0]);
+                    // Tell the real user what happened and how to reset the effective user.
+                    let err = error.make(403, 'Access denied');
+                    err.info =  `<p>You have changed the effective user to one with these properties:</p>` +
+                                `<div class="container"><pre class="bg-dark text-white rounded p-2">` +
+                                `uid = ${params.requested_uid}\n` +
+                                `role = ${params.requested_role}\n` +
+                                `mode = ${params.requested_mode}\n` +
+                                `</pre></div>` +
+                                `<p>This user does not have access to the course instance <code>${res.locals.course_instance.short_name}</code> (with long name "${res.locals.course_instance.long_name}").` +
+                                `<p>To reset the effective user, return to the <a href="${res.locals.homeUrl}">PrairieLearn home page</a>.</p>`;
+                    return next(err);
+                }
+
+                res.locals.authz_data.user = result.rows[0].user;
+                res.locals.authz_data.role = result.rows[0].role;
+                res.locals.authz_data.mode = result.rows[0].mode;
                 res.locals.req_date = result.rows[0].req_date;
-                // remove all course permissions if we are emulating another user
-                res.locals.authz_data.course_role = 'None';
-                res.locals.authz_data.has_course_permission_view = false;
-                res.locals.authz_data.has_course_permission_edit = false;
-                res.locals.authz_data.has_course_permission_own = false;
+
+                // Make sure that we never grant extra permissions
+                if (res.locals.authz_data.has_instructor_view) res.locals.authz_data.has_instructor_view = result.rows[0].has_instructor_view;
+                if (res.locals.authz_data.has_instructor_edit) res.locals.authz_data.has_instructor_edit = result.rows[0].has_instructor_edit;
+                if (res.locals.authz_data.has_course_permission_view) res.locals.authz_data.has_course_permission_view = result.rows[0].permissions_course.has_course_permission_view;
+                if (res.locals.authz_data.has_course_permission_edit) res.locals.authz_data.has_course_permission_edit = result.rows[0].permissions_course.has_course_permission_edit;
+                if (res.locals.authz_data.has_course_permission_own) res.locals.authz_data.has_course_permission_own = result.rows[0].permissions_course.has_course_permission_own;
```

Comment:
I'm more familiar with using logical-and to take the "minimum" of booleans. That is:
```
bool_var = bool_var && override_bool_var;
```

Reply:
haha, that would be a little nicer, wouldn't it